lib/rubocop/cop/performance/casecmp.rb

```diff
@@ -9,49 +9,82 @@ module Performance
       #
       # @example
       #   @bad
-      #   'abc'.downcase == 'ABC'
+      #   'ABC'.downcase == 'abc'
       #   'abc'.upcase.eql? 'ABC'
+      #   'abc' == 'ABC'.downcase
+      #   'ABC'.eql? 'abc'.upcase
+      #   'abc'.downcase == 'abc'.downcase
       #
       #   @good
       #   'abc'.casecmp('ABC').zero?
+      #   'abc'.casecmp('abc').zero?
+      #   'abc'.casecmp('ABC'.downcase).zero?
       class Casecmp < Cop
+        include IgnoredNode
+
         MSG = 'Use `casecmp` instead of `%s %s`.'.freeze
+        CASE_METHODS = [:downcase, :upcase].freeze
 
         def_node_matcher :downcase_eq, <<-END
-          (send $(send _ ${:downcase :upcase}) ${:== :eql? :!=} _)
+          (send $(send _ ${:downcase :upcase}) ${:== :eql? :!=} $_)
+        END
+
+        def_node_matcher :eq_downcase, <<-END
+          (send _ ${:== :eql? :!=} $(send _ ${:downcase :upcase}))
         END
 
         def on_send(node)
-          downcase_eq(node) do |send_downcase, case_method, eq_method|
-            range = node.loc.selector.join(send_downcase.loc.selector)
+          return if part_of_ignored_node?(node)
+
+          downcase_eq(node) do |send_downcase, case_method, eq_method, other|
+            *_, method = *other
+            if CASE_METHODS.include?(method)
```

Comment:
It is possible to inline `other` using _clojuresque_ destructuring:

```ruby
downcase_eq(node) do |send_downcase, case_method, eq_method, (*, method)|
  if CASE_METHODS.include?(method)
  ...
```

To me it's slightly more readable, but maybe it's just me. I don't have strong opinion about that.

What else bothers me a bit is the name of the variable `method`. It is not necessary a method, is it? Maybe it should be called `maybe_method`?

Or alternatively, we can leave `other` as is and extract `case_method_call?(expr)` method:

```ruby
downcase_eq(node) do |send_downcase, case_method, eq_method, other|
  if case_method_call?(other)
    range = ...
 ...

def case_method_call?(expr) # should it be def_node_matcher?
  CASE_METHODS.include?(expr.last)
end
```